File: src/policy.ts
/**
 * Consent policy administration. This is the gate that is actually
 * enforced.
 *
 *   npm run policy          # show what is in force
 *   npm run policy:allow    # permit the caller to read, time-boxed
 *   npm run policy:deny     # withdraw consent
 *
 * The delegation grant does not gate this contract, because the
 * platform's enforcement point is egress and this contract makes no
 * outbound call. So the contract checks the policy itself, against
 * values the node mints: the calling identity from tenant context, and
 * the cluster-pinned clock. A caller cannot claim to be someone else
 * and cannot move an expiry.
 *
 * Only the tenant identity may read or change the policy, checked the
 * same way inside the enclave.
 */
import { getContractVersion, getNodeUrl } from "@terminal3/t3n-sdk";
import {
  CONTRACT_TAIL,
  DECLARED_TENANT_DID,
  GRANT_TTL_SECS,
} from "./config.js";
import { openCaller } from "./caller.js";
import { canonicalName, openTenantSession } from "./session.js";

/** Functions consent covers when permitted. Read-only, deliberately. */
export const ALLOWED_FUNCTIONS = ["vault-read", "audit-list"];

/** Deliberately outside consent, so the gate is observably real. */
export const WITHHELD_FUNCTIONS = ["vault-put"];

interface Delegation {
  from: string;
  to: string;
  functions: string[];
  valid_until_secs: number | null;
}

interface PolicyView {
  version: number;
  allowed_callers: string[];
  allowed_functions: string[];
  valid_until_secs: number | null;
  delegations?: Delegation[];
  now_secs: number;
  expired: boolean;
}

interface PolicySetResult {
  version: number;
  allowed_callers: string[];
  allowed_functions: string[];
  valid_until_secs: number | null;
  audit_key: string;
}

/** Strip the `did:t3n:` prefix; the policy stores the 40 hex body. */
function didBody(did: string): string {
  return did.startsWith("did:t3n:") ? did.slice("did:t3n:".length) : did;
}

function describe(view: PolicyView): void {
  console.log(`policy version ${view.version}`);
  if (view.version === 0) {
    console.log(`  nothing in force, so every call is denied`);
    return;
  }
  console.log(
    `  callers   ${view.allowed_callers.length > 0 ? view.allowed_callers.join(", ") : "(none, so everyone is denied)"}`,
  );
  console.log(
    `  functions ${view.allowed_functions.length > 0 ? view.allowed_functions.join(", ") : "(none)"}`,
  );
  if (view.valid_until_secs === null) {
    console.log(`  expiry    none`);
  } else {
    const left = view.valid_until_secs - view.now_secs;
    console.log(
      `  expiry    ${new Date(view.valid_until_secs * 1000).toISOString()}` +
        ` (${view.expired ? "lapsed" : `${left}s left`})`,
    );
  }
  console.log(`  cluster time ${new Date(view.now_secs * 1000).toISOString()}`);
  const chain = view.delegations ?? [];
  if (chain.length > 0) {
    console.log(`  chain`);
    for (const d of chain) {
      const until = d.valid_until_secs === null ? "no expiry" : new Date(d.valid_until_secs * 1000).toISOString();
      console.log(`    ${d.from.slice(0, 8)}... -> ${d.to.slice(0, 8)}...  [${d.functions.join(", ")}]  until ${until}`);
    }
    console.log(`    each hop holds the intersection with what its delegator still holds`);
  }
}

async function main() {
  if (!DECLARED_TENANT_DID) {
    throw new Error("DID is not set in .env. It names the tenant that owns the contract.");
  }

  const action = (process.argv[2] ?? "show").toLowerCase();
  const tenant = await openTenantSession();
  const contractName = canonicalName(DECLARED_TENANT_DID, CONTRACT_TAIL);
  const version = await getContractVersion(getNodeUrl(), contractName);

  const call = <T,>(functionName: string, input: unknown) =>
    tenant.client.executeAndDecode<T>({
      contract_id: contractName,
      contract_version: version,
      function_name: functionName,
      pii_did: tenant.did,
      input,
    });

  console.log(`contract ${contractName}@${version}\n`);

  if (action === "show") {
    describe(await call<PolicyView>("policy-get", {}));
    console.log(`\nchange it with: npm run policy:allow | npm run policy:deny`);
    return;
  }

  if (action === "allow") {
    const caller = await openCaller();
    const until = Math.floor(Date.now() / 1000) + GRANT_TTL_SECS;
    const result = await call<PolicySetResult>("policy-set", {
      allowed_callers: [didBody(caller.did)],
      allowed_functions: ALLOWED_FUNCTIONS,
      valid_until_secs: until,
    });
    console.log(`consent granted, policy version ${result.version}`);
    console.log(`  caller    ${caller.did}   (${caller.role})`);
    console.log(`  functions ${result.allowed_functions.join(", ")}`);
    console.log(`  withheld  ${WITHHELD_FUNCTIONS.join(", ")}`);
    console.log(`  expires   ${new Date(until * 1000).toISOString()} (${GRANT_TTL_SECS}s)`);
    console.log(`  audit_key ${result.audit_key}`);
    console.log(
      `\nThe change is itself in the audit trail, so widening permission is\n` +
        `as visible as using it.`,
    );
    console.log(`\nnext: npm run invoke`);
    return;
  }

  if (action === "deny") {
    // An empty caller list is the honest encoding of withdrawn consent:
    // the document still exists and still has a version, so the trail
    // shows a deliberate withdrawal rather than an absent policy.
    const result = await call<PolicySetResult>("policy-set", {
      allowed_callers: [],
      allowed_functions: [],
      valid_until_secs: null,
    });
    console.log(`consent withdrawn, policy version ${result.version}`);
    console.log(`  audit_key ${result.audit_key}`);
    console.log(
      `\nEvery gated call now fails inside the enclave, with no redeploy.\n` +
        `Verify with: npm run invoke`,
    );
    return;
  }

  throw new Error(`unknown action "${action}". Use show, allow or deny.`);
}

main().catch((error: unknown) => {
  console.error(`\npolicy failed: ${error instanceof Error ? error.message : String(error)}`);
  process.exit(1);
});

File: contract/src/policy.rs
//! Consent policy, enforced by this contract inside the enclave.
//!
//! # Why the contract decides, rather than the delegation layer
//!
//! The platform's enforcement point for a tenant contract is egress: a
//! delegation grant names functions, but the reference implementation
//! revokes access by clearing allowed hosts and leaves the function
//! list populated, because the function still runs and only its
//! outbound call is denied. This contract makes no outbound call, so
//! there is nothing there for the platform to deny. Verified against
//! testnet: with the grant fully revoked, reads still succeeded.
//!
//! So the gate has to be a comparison this code makes.
//!
//! # Why that is enforcement and not merely intent
//!
//! Every value the decision rests on is minted by the node and cannot
//! be influenced by the caller:
//!
//!   * the calling identity comes from `tenant-context`, never from the
//!     request body, so a caller cannot claim to be someone else;
//!   * the clock is the cluster-pinned timestamp, not wall clock, so a
//!     caller cannot move an expiry.
//!
//! # What it does not claim
//!
//! Whoever can write the tenant's maps can rewrite this policy. The
//! guarantee is narrower and still worth having: a call cannot reach the
//! data without passing this check, and cannot pass it without being
//! recorded.

use serde::{Deserialize, Serialize};

/// The consent document, stored as one JSON value in the `policy` map.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Policy {
    /// Hex-encoded caller DIDs permitted to call at all. Empty denies
    /// everyone, which is the correct reading of "consent withdrawn"
    /// rather than an error.
    #[serde(default)]
    pub allowed_callers: Vec<String>,
    /// Function names those callers may invoke.
    #[serde(default)]
    pub allowed_functions: Vec<String>,
    /// When permission lapses. Absent means no expiry.
    #[serde(default)]
    pub valid_until_secs: Option<u64>,
    /// Bumped on each write and recorded in the audit trail, so a
    /// widening of permission is as visible as a use of it.
    #[serde(default)]
    pub version: u32,
}

impl Policy {
    /// The posture before any policy is written: deny everything.
    ///
    /// A missing document must not mean "allow", or forgetting to
    /// provision the map would silently disable the gate.
    pub fn deny_all() -> Self {
        Self {
            allowed_callers: Vec::new(),
            allowed_functions: Vec::new(),
            valid_until_secs: None,
            version: 0,
        }
    }
}

/// Why a call was refused. Carried into the response and the audit entry.
#[derive(Debug, PartialEq, Eq)]
pub enum Denial {
    NoPolicy,
    CallerNotListed,
    FunctionNotListed,
    Expired { valid_until_secs: u64, now_secs: u64 },
}

impl Denial {
    pub fn reason(&self) -> String {
        match self {
            Denial::NoPolicy => "no consent policy is in force".to_string(),
            Denial::CallerNotListed => "this caller is not permitted by the consent policy".to_string(),
            Denial::FunctionNotListed => {
                "the consent policy does not permit this function".to_string()
            }
            Denial::Expired {
                valid_until_secs,
                now_secs,
            } => format!(
                "consent lapsed at {valid_until_secs}, cluster time is {now_secs}"
            ),
        }
    }
}

/// Evaluate a call against the policy.
///
/// Checks run cheapest-first and refuse before anything reads data. The
/// order also decides which reason a caller learns, so it goes from the
/// least informative to the most: that a caller is unlisted reveals less
/// than which functions exist.
pub fn evaluate(
    policy: &Policy,
    caller_hex: &str,
    function: &str,
    now_secs: u64,
) -> Result<(), Denial> {
    if policy.version == 0 && policy.allowed_callers.is_empty() {
        return Err(Denial::NoPolicy);
    }

    // Compare case-insensitively: a DID's identity is the bytes it
    // decodes to, not the spelling of its hex.
    let listed = policy
        .allowed_callers
        .iter()
        .any(|c| c.eq_ignore_ascii_case(caller_hex));
    if !listed {
        return Err(Denial::CallerNotListed);
    }

    if !policy.allowed_functions.iter().any(|f| f == function) {
        return Err(Denial::FunctionNotListed);
    }

    if let Some(until) = policy.valid_until_secs {
        if now_secs >= until {
            return Err(Denial::Expired {
                valid_until_secs: until,
                now_secs,
            });
        }
    }

    Ok(())
}

pub const POLICY_KEY: &[u8] = b"consent";

#[cfg(target_arch = "wasm32")]
use crate::host::interfaces::kv_store;

/// Load the policy. An absent or unreadable document denies everything
/// rather than failing the call, so the gate stays closed either way and
/// the attempt is still recorded.
#[cfg(target_arch = "wasm32")]
pub fn load() -> Result<Policy, String> {
    let map = crate::map_name(crate::POLICY_TAIL);
    let found = kv_store::get(&map, POLICY_KEY).map_err(|e| format!("policy read {map}: {e}"))?;
    match found {
        None => Ok(Policy::deny_all()),
        Some(bytes) => Ok(serde_json::from_slice(&bytes).unwrap_or_else(|_| Policy::deny_all())),
    }
}

#[cfg(target_arch = "wasm32")]
pub fn store(policy: &Policy) -> Result<(), String> {
    let map = crate::map_name(crate::POLICY_TAIL);
    let value = serde_json::to_vec(policy).map_err(|e| e.to_string())?;
    kv_store::put(&map, POLICY_KEY, &value).map_err(|e| format!("policy write {map}: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn policy() -> Policy {
        Policy {
            allowed_callers: vec!["abe8e6dc".to_string()],
            allowed_functions: vec!["vault-read".to_string()],
            valid_until_secs: Some(1000),
            version: 1,
        }
    }

    #[test]
    fn allows_a_listed_caller_and_function_before_expiry() {
        assert!(evaluate(&policy(), "abe8e6dc", "vault-read", 999).is_ok());
    }

    #[test]
    fn a_missing_policy_denies_rather_than_allows() {
        // Forgetting to provision the map must not disable the gate.
        let err = evaluate(&Policy::deny_all(), "abe8e6dc", "vault-read", 1).unwrap_err();
        assert_eq!(err, Denial::NoPolicy);
    }

    #[test]
    fn denies_an_unlisted_caller() {
        let err = evaluate(&policy(), "deadbeef", "vault-read", 1).unwrap_err();
        assert_eq!(err, Denial::CallerNotListed);
    }

    #[test]
    fn denies_a_function_the_policy_does_not_name() {
        // This is the check the delegation grant failed to make.
        let err = evaluate(&policy(), "abe8e6dc", "vault-put", 1).unwrap_err();
        assert_eq!(err, Denial::FunctionNotListed);
    }

    #[test]
    fn denies_once_consent_has_lapsed() {
        let err = evaluate(&policy(), "abe8e6dc", "vault-read", 1000).unwrap_err();
        assert!(matches!(err, Denial::Expired { .. }));
        // The boundary is exclusive: at exactly the expiry it is over.
        assert!(evaluate(&policy(), "abe8e6dc", "vault-read", 1001).is_err());
    }

    #[test]
    fn treats_hex_case_as_the_same_identity() {
        // A DID's identity is the bytes it decodes to, not its spelling.
        assert!(evaluate(&policy(), "ABE8E6DC", "vault-read", 1).is_ok());
    }

    #[test]
    fn an_emptied_caller_list_denies_everyone() {
        // This is what withdrawing consent looks like on the wire.
        let mut p = policy();
        p.allowed_callers.clear();
        p.version = 2;
        let err = evaluate(&p, "abe8e6dc", "vault-read", 1).unwrap_err();
        assert_eq!(err, Denial::CallerNotListed);
    }

    #[test]
    fn no_expiry_means_it_does_not_lapse() {
        let mut p = policy();
        p.valid_until_secs = None;
        assert!(evaluate(&p, "abe8e6dc", "vault-read", u64::MAX).is_ok());
    }
}
